src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs

```diff
@@ -12,14 +12,434 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics
+    }
+
+    public Enumerator GetEnumerator() => new Enumerator();
+}
+
+class C
+{
+    void M()
+    {
+        foreach (ref int x in new E())
+        {
+            int y = 0;
+            x = ref y;
+        }
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (18,18): error CS1073: Unexpected token 'ref'
+                //         foreach (ref int x in new E())
+                Diagnostic(ErrorCode.ERR_UnexpectedToken, "ref").WithArguments("ref").WithLocation(18, 18),
+                // (21,13): error CS8355: The left-hand side of a ref assignment must be a ref local or parameter.
+                //             x = ref y;
+                Diagnostic(ErrorCode.ERR_RefLocalOrParamExpected, "x").WithArguments("x").WithLocation(21, 13));
+        }
+
+        [Fact]
+        public void RefReassignNarrowLifetime()
+        {
+            var comp = CreateStandardCompilation(@"
+class C
+{
+    int _x = 0;
+    void M()
+    {
+        int y = 0;
+        ref int rx = ref _x;
+        rx = ref y;
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (9,9): error CS8356: Cannot ref-assign 'y' to 'rx' because 'y' has a narrower escape scope than 'rx'.
+                //         rx = ref y;
+                Diagnostic(ErrorCode.ERR_RefAssignNarrower, "rx = ref y").WithArguments("rx", "y").WithLocation(9, 9));
+        }
+
+        [Fact]
+        public void RefReassignRangeVar()
+        {
+            var comp = CreateStandardCompilation(@"
+using System;
+using System.Linq;
+class C
+{
+    void M()
+    {
+        _ = from c in ""test"" select (Action)(() =>
+        {
+            int x = 0;
+            ref int rx = ref x;
+            rx = ref c;
+            c = ref x;
+        });
+    }
+}", references: new[] { LinqAssemblyRef });
+            comp.VerifyDiagnostics(
+                // (12,22): error CS1939: Cannot pass the range variable 'c' as an out or ref parameter
+                //             rx = ref c;
+                Diagnostic(ErrorCode.ERR_QueryOutRefRangeVariable, "c").WithArguments("c").WithLocation(12, 22),
+                // (13,13): error CS8355: The left-hand side of a ref assignment must be a ref local or parameter.
+                //             c = ref x;
+                Diagnostic(ErrorCode.ERR_RefLocalOrParamExpected, "c").WithArguments("c").WithLocation(13, 13));
+        }
+
+        [Fact]
+        public void RefReassignOutDefiniteAssignment()
+        {
+            var comp = CreateStandardCompilation(@"
+class C
+{
+    static int y = 0;
+    void M(out int x)
+    {
+        x = ref y;
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (5,10): error CS0177: The out parameter 'x' must be assigned to before control leaves the current method
+                //     void M(out int x)
+                Diagnostic(ErrorCode.ERR_ParamUnassigned, "M").WithArguments("x").WithLocation(5, 10));
```

Comment:
Just to confirm, if you did `x = 0;` before the method ends, you'd still get the same error, right?